Results screen additions and 1 ms local offset adjustment with Ctrl

YaLTeR opened the pull request:
![image](https://user-images.githubusercontent.com/1794388/61179023-e8caad00-a5ae-11e9-98e9-7f4bb3e6ba1a.png)

Things to consider:
- I would guess that the main button to be used here is the local offset adjustment one, so it might make sense to emphasize it somehow to make it easier to click. Perhaps put the buttons horizontally, below the divider line?
- You want to click only one of these buttons once, and only when your global or local offset equals to that which you used when playing the score. Perhaps we should store the global and local offset used in the score and then base the new offset calculation off of that instead of the current offset? This way it will always set the offset to the correct value.
- Not sure what the best place would be to put `FilterHitStats()`. I think there needs to be some place in `ResultsScreen` where all this data is computed once for all sub-components to use it?
- Maybe "average" sounds better than "mean"?
- "Adjust Map Offset" instead of "Adjust Local Offset"?
- Thinking about it, the LN release judgement window scale should probably _not_ be applied when computing mean and stddev.

Fixes ...uhh, I was sure there were at least 2 issues about this stuff, but I guess not. I believe this offers a better solution than #949 though.

YaLTeR commented:
Todo offset adjustment should be mean divided by rate not just mean

Swan commented:
* >Perhaps put the buttons horizontally, below the divider line? - I would say place the buttons below the graph horizontally. You can make the buttons bigger. As a side note, we'll want to be using `SpriteTextBitmap` for text. `BorderedTextButton` uses the old SpriteText
* Sure
* Quaver.API possibly?
* Average
* It's called local offset everywhere else. No need to change it
* Why?

YaLTeR commented:
> * Why?

Imagine if autoplay is playing an LN map with a "wrong offset", so all hits will be, say, 10 ms off. Currently mean will show less than 10 ms because half of the hits are LN releases, so they will be scaled 1.5x down. However, what we really want here is to show exactly 10 ms and to adjust the offset by 10 ms.

YaLTeR commented:
Also could you comment on storing the offsets in the score (question 2)?

YaLTeR commented:
![image](https://user-images.githubusercontent.com/1794388/61191209-57b20f80-a65c-11e9-8df1-3971a75b5f15.png)

YaLTeR commented:
TODO: the stuff done in `GetScoreProcessor()` should be done once and cached.

Swan commented on Quaver.Shared/Screens/Result/UI/ResultScoreContainer.cs:
```diff
@@ -69,6 +73,11 @@ public class ResultScoreContainer : Sprite
+        /// <summary>
+        ///     The horizontal divider line under the statistics key/value items.
+        /// </summary>
+        public Sprite StatisticsKeyValueItemDividerLine { get; private set; }
```
Instead of having this, just extend `ResultKeyValueItemDividerLine` to the full width

Swan commented on Quaver.Shared/Screens/Result/UI/ResultScoreContainer.cs:
```diff
@@ -270,6 +280,125 @@ private void CreateKeyValueItems()
+            var processor = Screen.GetScoreProcessor();
+            var largestHitWindow = processor.JudgementWindow.Values.Max();
+
+            double mean = 0.0, standardDeviation = 0.0;
```
```cs
double mean;
double standardDeviation;
```

Swan commented on Quaver.Shared/Screens/Result/UI/ResultScoreContainer.cs:
```diff
@@ -270,6 +280,125 @@ private void CreateKeyValueItems()
+            var buttonPadding = 15;
+            var buttonWidth = (availableWidth - buttonPadding * 3) / 2;
+
+            var localOffsetButton = new BorderedTextButton("Adjust Local Offset", Colors.MainAccent,
```
You shouldn't be creating buttons here if the method is supposed to be creating KeyValue items

Swan replied:
`BorderedTextButton` uses the older text. It's obsolete and should never be used. You can either create a new button or modify `TextButton` to use `SpriteTextBitmap`

YaLTeR replied:
> `BorderedTextButton` uses the older text. It's obsolete and should never be used. You can either create a new button or modify `TextButton` to use `SpriteTextBitmap`

Sounds out of scope for this PR. Bordered text buttons are used in a bunch of places still, so a subsequent PR could change them to use the new text.

Swan commented on Quaver.Shared/Screens/Result/UI/ResultScoreContainer.cs:
```diff
@@ -270,6 +280,125 @@ private void CreateKeyValueItems()
+                    // score was played on.
+                    var change = mean * ModHelper.GetRateFromMods(processor.Mods);
+                    ResultScreen.Map.LocalOffset =
+                        (int) Math.Round(ResultScreen.Map.LocalOffset - change);
```
Put on the same line

Swan replied:
If `change` isn't being used anywhere else, there's no need to make a local variable for it

YaLTeR replied:
The local variable makes the expression more readable.

Swan commented on Quaver.Shared/Screens/Result/UI/ResultScoreContainer.cs:
```diff
@@ -270,6 +280,125 @@ private void CreateKeyValueItems()
+
+                    MapDatabaseCache.UpdateMap(ResultScreen.Map);
+                    NotificationManager.Show(NotificationLevel.Success,
+                        $"Local offset was set to {ResultScreen.Map.LocalOffset} ms.");
```
Put on the same line (and everything else similar to this)

Swan commented on Quaver.Shared/Screens/Result/UI/ResultScoreContainer.cs:
```diff
@@ -270,6 +280,125 @@ private void CreateKeyValueItems()
+                    FontSize = 9
+                }
+            };
+            var globalOffsetButton = new BorderedTextButton("Adjust Global Offset", Colors.MainAccent,
```
The one issue I have with these buttons are that there should probably be some sort of confirmation button making the user aware of what their new offset will be and if they'd like to proceed. It can be in the form of a `ConfirmCancelDialog`. If not this, then the option to undo should be there, but I prefer the former.

Swan replied:
Also, it's probably only worth creating these buttons only after gameplay. You wouldn't want to always display it (especially when clicking on an online leaderboard score or watching a replay)

YaLTeR replied:
There is value in having them on local scores: I used them a couple of times to adjust the offset from a score that I had before this was implemented, and also a couple of times when I forgot to click the button before escaping out of the gameplay score.

Swan replied:
There are some instances where you don't have access to these stats such as online scores and multiplayer (for other players). Hiding it when `Score.IsOnline` is true should do the trick.

YaLTeR commented:
Rebased and addressed all comments. Now depends on https://github.com/Quaver/Quaver.API/pull/67.

I'd really prefer _not_ to do the `ScoreProcessor` refactor outlined in the TODO in this PR.

Swan commented:
@YaLTeR  Quaver.API submodule can also be updated
